### ISSUE TITLE
support for custom indexes

### ISSUE THREAD

**Issue by wanderer** [balderdashy/waterline#109]
It would be nice to be able to specify custom indexes. so for example `index: "customIndex"`  the string `customIndex` would get passed along to the adapter.

If the index was a bool then the current behavior would hold.

**Comment by andyhu** [comment]
+1 for composite keys

**Comment by dmarcelino** [comment]
+1 for composite keys and I'm OK with @wanderer's suggestion [above](https://github.com/balderdashy/waterline/issues/109#issuecomment-23899495).

**Comment by devinivy** [comment]
Yep, the format that @wanderer laid out covers many use-cases, including important indexing cases as in databases like DynamoDB.  It would be great to implement the list of indices optionally as an object instead of an array so that the indices can be given names (their key in the object).  Or perhaps allow an array for unnamed indices and an object for named indices.  I think this project will necessarily have to support index definitions in various levels of explicitness (perhaps the two that @wanderer mentioned, plus allowing named indices).

Simple
```
indexes:[
//a composite index
{
   attributes:  ['attribute1', 'attribute2' ...]
   unique: boolean
}
]
```
More explicit
```
indexes:[
//a composite index
{
   attributes:  [
        {
            name: 'attribute1',
            type: '2dsphere'
        },
        {
            name: 'attribute1',
            type: 'hash'
        }
   ]
   unique: boolean
}
]
```
Most explicit
```
indexes:{
    //a composite index
    'indexName': {
        attributes:  [
            {
                name: 'attribute1',
                type: '2dsphere'
            },
            {
                name: 'attribute1',
                type: 'hash'
            }
       ]
       unique: boolean
    }
}
```

**Comment by calidion** [comment]
+1 for composite keys

**Comment by franc23** [comment]
+1 for composite keys

**Comment by byCedric** [comment]
+1 for the ability to use composite keys

**Comment by syshock1337** [comment]
Data integrity should be maintained at the lowest possible level

+1 for composite keys

**Comment by franc23** [comment]
@particlebanana or any collaborator: I have hacked into the mysql adapter to get composite keys to work with get/put calls (post not being an issue). I make use of the unique attribute to modify the adapter.js methods and a few other methods in 6 files total. 
Maybe wrong in my approach, but thought I'd share this with the team to see if it helps. Would be happy to help as much as I can if someone can guide me on any waterline standards, etc.

Uploaded the files with appropriate folder paths here  -http://www.megafileupload.com/K1/node_modules.zip

**Comment by albertkim** [comment]
I've hacked around this issue by having big beforeCreate functions in my models involving composite keys. Thankfully none of my tables require more than 2 composites, but if there are any more I don't think this method will hold up. Here's a +1 from me.

I look forward to seeing how this gets implemented.

**Comment by elexx** [comment]
I am also struggling with not having composite keys. +1 from me.

**Comment by Salakar** [comment]
I created a sails hook to give advanced indexing options for models that use the sails-mongo adapter.

Supports all mongo indexing options.

https://www.npmjs.com/package/sails-hook-mongoat

**Comment by unodelostrece** [comment]
+1

**Comment by BLamy** [comment]
This is really throwing a huge wrench in my database design.

**Comment by devinivy** [comment]
@BLamy you can always manually create indexes or use migration tools to do so.

**Comment by BLamy** [comment]
@devinivy It's not just the indexes. It's the primary keys. Is it possible to create a composite primary key using migration tools or do it behind the scenes and not break sails-mysql?

**Comment by devinivy** [comment]
Yes– if `migrate` is set to `safe` or `create` then you should be fine.  `alter` and `drop` are not suitable for production anyway :+1:

**Comment by sailsbot** [comment]
Thanks for posting, @wanderer.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by dmarcelino** [comment]
Reopening as this one has been picked for `WL_Next`, likely to be renamed 1.2.

**Comment by sailsbot** [comment]
Thanks for posting, @wanderer.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by LeonardoGentile** [comment]
+1
So being closed by sailsbot means it is the end of this FR? 
Any news on this from human beings?

**Comment by RWOverdijk** [comment]
I wish I could reopen but I can't. @devinivy

**Comment by devinivy** [comment]
Reopening, just until someone PRs this into the [roadmap](https://github.com/balderdashy/waterline/blob/master/ROADMAP.md).  

@LeonardoGentile mind making a PR to add this feature into the roadmap?

**Comment by LeonardoGentile** [comment]
@devinivy :+1: done

**Comment by calidion** [comment]
:+1:

**Comment by arcseldon** [comment]
@particlebanana,  @CWyrtzen, @albertpeiro, @RWOverdijk - does anyone have any further information regarding this feature request?  It continues to generate frequent votes, and it is unclear (to me at least) whether any of the core committers / sails / waterline community are working on a solution. Has the conversation moved elsewhere (IRC channels etc) or is this still the correct issue to be expressing opinions / providing updates etc?  Thanks for any clarifications, and please keep up the good work on these excellent libraries which is much appreciated.

**Comment by devinivy** [comment]
@arcseldon this is now in the roadmap.  I don't think it's anyone's duty right _now_ to add the feature, but it's scheduled for the future.  In the meantime you can support custom indices on the adapter level by exploiting the permissiveness of waterline models– see https://github.com/gadelkareem/sails-dynamodb for an example.  If you'd like to help implement a solution, feel free to reach out!

**Comment by particlebanana** [comment]
@arcseldon it's on the roadmap. No work has started on it but a PR implementing it would be accepted!

Ideally we could move all the migrations and schema building to a standalone library eventually.

**Comment by calidion** [comment]
seems you are eventually encountered this problem:)

**Comment by arcseldon** [comment]
@devinivy, @particlebanana - thanks for your prompt replies. Good to know it's on the roadmap! Working overtime on other projects right now (who isn't right?) but will surely take a look if time permits next year.

**Comment by icodeforlove** [comment]
:+1:

**Comment by gadidev** [comment]
I want this.   This discussion is over 2 years old.  I hope its coming soon.

**Comment by tjwebb** [comment]
> This discussion is over 2 years old. I hope its coming soon.

This statement is inconsistent with itself :)  If it hasn't happened in 2 years, it'll probably never happen. See also: https://github.com/balderdashy/sails/issues/3429#issuecomment-165004024

**Comment by gadidev** [comment]
yeah.  I switched my project from Laravel to Sails about 2 months ago.  I'm starting to wonder if I made a mistake.

**Comment by RWOverdijk** [comment]
@gadidev I have to admit it has been quiet for a bit. But rest assured, things are being picked up again. I for one am still here :) I too think this is a useful feature (which I'll be needing really soon).

I think we have two options here:

**Option 1:** Extend the behavior of index. Allow for an index identifier to be set, and group indexes by that identifier. E.g. 
```javascript
{
  gender: {type: 'string', index: {name: 'idx_gender_age'}},
  age: {type: 'integer', index: {name: 'idx_gender_age'}}
};
```

**Option 2:** Add a `indexes` property, holding an object,  where you specify the index name as key, and  an array or string value as the columns to index, like so:
```javascript
{
  indexes: {idx_gender_age: ['gender', 'age']},
  attributes: {
    gender: 'string',
    age: 'integer'
  }
}
```

Implementing support for this shouldn't be too tricky, but isn't BC either way (if someone is using the properties right now, or incorrectly using index). Also this would have to be implemented on the adapter level. Thoughts?

**Comment by gadidev** [comment]
I too prefer #2.     #1 seems clunky `index: {name: 'idx_gender_age'}` is repeated.
Option #2 seems more natural, since you could have multiple indexes with the same attributes.
I was planning to use this prevent the same combination of fields (ie gener+age) from being inserted.  ie, unique property.  Thanks for taking this on.

**Comment by RWOverdijk** [comment]
@gadidev Good one! We'd also need support for index type.

**Comment by AzriAhmad** [comment]
+1 for composite keys

**Comment by Tronix117** [comment]
+1

**Comment by vladfr** [comment]
+1 

Was this added? because all comments point to here, but this issue is closed.

**Comment by Tronix117** [comment]
I agree this is a really important feature, which,  among others, Waterline doesn't support yet but should. On our projects, we are currently discussing about replacing Waterline with Knex or Sequelize.

**Comment by devinivy** [comment]
It's perfectly reasonable to use Waterline as an ORM and Knex to help
manage migrations.

**Comment by Tronix117** [comment]
@devinivy Totally using Knex for migrations right now, when I said Knex, I wanted to say Bookshelfjs, the ORM based on it. We didn't dig too much, but it seems worth discussions.
However Sequelize totaly handle indexes or composite primary keys, which Waterline does not.

**Comment by sio-iago** [comment]
Feels pretty bad when you start working with JavaScript, get crazy excited and bump into an issue like this in the middle of a project.

I'm not here to be a fanboy, but I'd suggest you guys to use Hibernate/Doctrine as a benchmark to Waterline. Some really important features (and composite key is just one of many I needed) were not developed.

The framework is well developed and easy to use, but it's pending on some really important features guys.

**Comment by RWOverdijk** [comment]
@sio-iago Perhaps not the place to mention this, but you might want to star [wetland](https://github.com/SpoonX/wetland). We're currently actively developing it so it should come out soon(ish). It'll first be a drop in replacement (more or less) for waterline. The reason I mention it is because it's closely modeled to the jpa-spec (570 pages of goodness), borrowing principles from doctrine, and that you mentioned doctrine. I thought, perhaps you might want to weigh in on the development (or brainstorm a bit). If so, you can find me [here](https://gitter.im/SpoonX/Dev).

**Comment by jvanalst** [comment]
Ran into this trying to hook sails up to a legacy database. Not sure what I'm going to do now.

**Comment by sio-iago** [comment]
@jvanalst either cry or use knex/bookshelf.

**Comment by arunramachandran15** [comment]
Is this issue fixed ?

**Comment by Tronix117** [comment]
@arunramachandran15 Like previous answer, it's not fixed and I don't think it will be before a long time. Anyway it is not a major issue, since this notion of indexes is only used to automaticaly change the schemas of your database in your DEV environment.

If you need good indexes, it means you expect to have a lot of data, thus you can not rely on waterline mecanism to upgrade your database (I explain why at the end of this message).

You should use sails-migration (which uses Knex) to create a file each time you need to change the schema of your database. You can do pretty much what you want, as well as multi key indexes. It's the recommended solution, and plus it allow you a great control over your schema by giving you the possibility to rollback to a previous migration.

---
You should always have `migrate: "safe"`, otherwise each time you modify a column or change something in the model sails will have the following behavior (even with `migrate: alter` on mysql) : 

1. Dump the corresponding table in memory
2. Drop the table
3. Recreate the table following the specs in the model file
4. Restore the dump from memory in the newly created table (datas that does not fit anymore are lost)

It may be great in development, but totaly not viable once you have a lot of datas in production.

**Comment by jvanalst** [comment]
The solution we ended up settling on for composite primary keys was to add new id row to the database that was a combination of id1 & id2 (e.g. id = id1*10000 + id2). It's hackney'd and very specific to our situation (since id2 will never realistically be above 9999). But it gets us up and running quickly with legacy data.

**Comment by AjkP** [comment]
+1 on adding support for this...
3.5 years of conversation on this and no definite answer...

**Comment by rassulcrowe** [comment]
+1. I am building a real time chat app and store all the messages of the users in database. So my db has a table "Conversations" with two fields userA and userB(users having conversation with each other) and this pair should be unique, pls could anyone advice how to solve this in Waterline(the solution with direct constraints to db isn't so elegant)?

**Comment by Tronix117** [comment]
@rassulcrowe Like I said before, the solution using directly the DB with migrations (with knee) is elegant and very clean because waterline handles migrations really poorly and should only be used in safe mode. You can eventually add a custom validation to check the DB for duplicate on inserting.

**Comment by omidh28** [comment]
+1

**Comment by stokesbga** [comment]
If you're not using legacy db, the solution is to abandon REST and just use RPC. Who cares. Not me!

**Comment by yanlinaung30** [comment]
Conversation started 4 years and 8 months ago and no definite answer yet.

**Comment by arcseldon** [comment]
That's about how long ago i last used Sails.js or Waterline. They were decent open source projects back in the day, but getting traction on maintenance & support for valuable issues / feature gaps was never a good story.  Just setting expectations.

**Comment by mj-mehdizadeh** [comment]
+1
